pkg/scheduler/queue/query_component_capacity.go

```diff
@@ -0,0 +1,179 @@
+// which are currently in flight processing a query which requires that QueryComponent.
+//
+// The component name can indicate the usage of one or both of ingesters and store-gateways,
+// and will be indicated as exceeding the threshold if either ingesters or store-gateways
+// are currently in excess of the reserved capacity.
+//
+// The threshold for a QueryComponent's utilization of querier-worker connections
+// can only be exceeded by one QueryComponent at a time as long as targetReservedCapacity is < 0.5.
+// Therefore, one of the components will always be given the OK to dequeue queries for.
+//
+// Capacity reservation only occurs when the queue backlogged, where backlogged is defined as
+// (length of the query queue) >= (number of querier-worker connections waiting for a query).
+//
+// A QueryComponent's utilization is allowed to exceed the reserved capacity when the queue is not backlogged.
+// If an influx of queries then creates a backlog, this method will indicate to skip queries for the component.
+// As the inflight queries complete or fail, the component's utilization will naturally decrease.
+// This method will continue to indicate to skip queries for the component until it is back under the threshold.
+func (qcl *QueryComponentCapacity) ExceedsCapacityForComponentName(
+	name string, connectedWorkers *atomic.Int64, queueLen, waitingWorkers int,
+) (bool, QueryComponent) {
+	if waitingWorkers > queueLen {
+		// excess querier-worker capacity; no need to reserve any for now
+		return false, ""
+	}
+	if connectedWorkers.Load() <= 1 {
+		// corner case; cannot reserve capacity with only one worker available
+		return false, ""
+	}
+
+	// reserve at least one connection in case (connected workers) * (reserved capacity) is less than one
+	minReservedConnections := int64(
+		math.Ceil(
+			math.Max(qcl.targetReservedCapacity*float64(connectedWorkers.Load()), 1),
+		),
+	)
+
+	isIngester, isStoreGateway := queryComponentFlags(name)
+	if isIngester {
+		if connectedWorkers.Load()-(qcl.ingesterInflightRequests.Load()) <= minReservedConnections {
+			return true, Ingester
+		}
+	}
+	if isStoreGateway {
+		if connectedWorkers.Load()-(qcl.storeGatewayInflightRequests.Load()) <= minReservedConnections {
+			return true, StoreGateway
+		}
+	}
+	return false, ""
+}
+
+// IncrementForComponentName is called when a request is sent to a querier
+func (qcl *QueryComponentCapacity) IncrementForComponentName(expectedQueryComponent string) {
+	qcl.updateForComponentName(expectedQueryComponent, 1)
+}
+
+// DecrementForComponentName is called when a querier completes or fails a request
+func (qcl *QueryComponentCapacity) DecrementForComponentName(expectedQueryComponent string) {
+	qcl.updateForComponentName(expectedQueryComponent, -1)
+}
+
+func (qcl *QueryComponentCapacity) updateForComponentName(expectedQueryComponent string, increment int64) {
```

Comment:
I try to use `expectedQueryComponent` everywhere because it does not tell the querier which query component to call for, it just tries to predict it.

It should be very rare, but the threshold for query-ingesters-within or query-store-after could be crossed after this gets annotated in the front end and before they query is actually executed, making it inaccurate.

Reply:
Right, makes sense, thanks for clarifying.